lib/pbench/test/unit/server/database/test_server_config_db.py

```diff
@@ -0,0 +1,199 @@
+import pytest
+
+from pbench.server.database.models.server_config import (
+    ServerConfig,
+    ServerConfigBadKey,
+    ServerConfigBadValue,
+    ServerConfigDuplicate,
+    ServerConfigMissingKey,
+)
+
+
+class TestServerConfig:
+    def test_bad_key(self, db_session):
+        """Test server config parameter key validation"""
+        with pytest.raises(ServerConfigBadKey) as exc:
+            ServerConfig.create(key="not-a-key", value="no-no")
+        assert str(exc.value) == "Configuration key 'not-a-key' is unknown"
```

Comment:
*All* of our unit tests are pseudo-functional, because that's the template we started working from both against SQLAlchemy and against Flask. In theory I don't like it, and in practice it's just a total pain to start building up the library of mocks necessary to combat it.

So on the one hand I can't help feel offended that you're bringing that up now in this context when I'm just trying to add a new feature.

On another hand (perhaps one of several), that's how it'll always be, and we do need to start somewhere ... so you're not wrong. ☹️

Reply:
I certainly do not intend to offend, and I deeply apologize if I have done so.

The place I would like to get to is where all new code comes with unit tests in the PR.  So, the fact that this is a new feature means that it is exactly the thing which should be accompanied by unit tests.  (And, soon, we'll want new features to come with functional tests as well....)

And, I realize that it's way easier to use the previous database tests (and their existing infrastructure) to compose new tests, but we need to start the transition somehow, and this seems like a good place to do that.

As for building up a library of mocks, I'm hoping that that won't be necessary, in general.  Most of the mocks that we need for this code are very simple, to the point of nearly trivial (and they are probably best implemented as closures in the test functions, anyway, so that they can grab the values that they should be expecting to receive from the test function scope), so they don't need to be part of a library.  Yes, there might be things that we will want to reuse between tests or test modules, but we can build them as we need them and refactor to make them available.

The key is to get started....  🙂